Consume messages with multiple topics and custom timeout

mikemintz opened the pull request:
This simplifies a bit of code, and makes two bigger changes:

1. Implement `consumeNumberMessagesFromTopics`, which reads from multiple topics.

2. Add custom timeout in `consumeNumberMessagesFromTopics`, instead of hard-coding 5 seconds. This also counts the total consume time toward the timeout, rather than timing out if there is a 5 second period of silence. E.g. if a user calls this method with number=12 and timeout=5.seconds, and a producer produces one message every 4 seconds, the method will now throw TimeoutException after 5 seconds. The previous behavior would receive all messages in 60 seconds and not throw an exception, because there was at least one message to consume every 5 seconds.

Since change 2 is an actual behavior change, I can take it out of this PR if backwards compatibility is necessary.

My use case is running a complicated algorithm that publishes many messages to many topics, and I want to confirm that the right thing was published everywhere. I know my producer should be done after no more than 30 seconds. But if it fails, I don't want my test to stall for more than 30 seconds.

manub commented on embedded-kafka/src/main/scala/net/manub/embeddedkafka/EmbeddedKafka.scala:
```diff
@@ -304,96 +304,74 @@ sealed trait EmbeddedKafkaSupport {
+      if (messagesRead < number) {
+        throw new TimeoutException(s"Unable to retrieve $number message(s) from Kafka in $timeout")
+      }
+      messagesBuffers.map(kv => kv._1 -> kv._2.toList)
```
Please expand `kv` using `case (topic, messages)`

mikemintz commented:
I changed the `kv` to `case` syntax, and I added an `resetTimeoutOnEachMessage` option to preserve the old behavior as the default. If you prefer, I can split it out into two separate methods, but I thought adding a flag to the existing method might be nice to minimize code duplication.

manub commented:
Thanks, will release hopefully this weekend (I'll have limited access to my computer for the rest of the month)
